File: sdk/zig/gearbox.zig
//! gearbox.zig — OpenDoctrines mod ABI, Gearbox v1.0
//!
//! A transcription of `sdk/abi.json`. That file is the source of truth; if this
//! one disagrees with it, this one is a bug.
//!
//! Two directions:
//!   - IMPORTS (`raw.*`)  the host provides, your mod calls.
//!   - EXPORTS (`mod_*`)  your mod provides, the host calls. You write those.
//!
//! Imports live in WASM modules named `gearbox:<module>`, matching the
//! capability names in MANIFEST.json. You only receive imports for capabilities
//! you declared AND the user granted; an undeclared import is refused at load
//! time with a diagnostic, not at some later inconvenient moment.
//!
//! Only the imports you actually call are emitted. Zig analyses declarations
//! lazily and wasm-ld only keeps referenced undefined symbols, so merely
//! importing this file does not put `gearbox:assets` in your module. That is
//! what you want: an import you do not use would still have to be granted.
//!
//! Strings on the wire are (ptr, len) pairs of UTF-8 bytes in your linear
//! memory. They are NOT null-terminated. The host does not retain a pointer
//! into your memory after a call returns, and you must not retain one of its.
//!
//! Status: written against sdk/abi.json, not yet compiled.

pub const MAJOR: u32 = 1;
pub const MINOR: u32 = 0;

/// Returned by any call that yields a handle it could not produce.
pub const INVALID: u32 = 0xFFFF_FFFF;

/// Opaque handles. Valid only for the duration of the hook that produced them —
/// do not cache one across turns, it will be stale or reassigned. Store the
/// country's name or your own key instead.
pub const Country = u32;
pub const Province = u32;
pub const Panel = u32;

pub const LogLevel = enum(u32) {
    trace = 0,
    info = 1,
    warn = 2,
    /// `error` is a Zig keyword, so the ABI's ERROR is spelled `err` here.
    err = 3,
};

pub const Platform = enum(u8) {
    unknown = 0,
    windows = 1,
    macos = 2,
    linux = 3,
    web = 4,
    _,
};

/// Layout is ABI. Fields are only ever appended, never reordered or resized;
/// `size` is what lets an older mod stay safe against a newer host.
///
/// Every field defaults to 0 so you can write `var e: Env = .{};` — but use
/// `env()` below, which fills `size` for you before calling the host.
pub const Env = extern struct {
    /// You write @sizeOf(Env) here before calling `raw.env`.
    size: u32 = 0,
    gearbox_major: u32 = 0,
    gearbox_minor: u32 = 0,
    /// Packed: (major << 16) | (minor << 8) | patch.
    host_version: u32 = 0,
    /// A `Platform` value. Kept as u8 because the host may add platforms.
    platform: u8 = 0,
    /// 1 under Emscripten. Fuel is NOT enforced there.
    is_web: u8 = 0,
    /// 1 when there is no renderer. Every UI import silently no-ops.
    is_headless: u8 = 0,
    reserved0: u8 = 0,
    /// 0 when headless.
    screen_w: u32 = 0,
    screen_h: u32 = 0,
};

comptime {
    // The four u8 fields pack into one word; wasm32 uses 4-byte alignment for
    // u32, so this must come out at exactly the 28 bytes abi.json specifies.
    if (@sizeOf(Env) != 28) @compileError("Env must be 28 bytes to match the ABI");
}

// ============================================================== imports ====
//
// The wire ABI, one Zig declaration per entry in abi.json's "imports" array.
// Pointer parameters that the ABI allows to be null are `?[*]`; the rest are
// plain many-item pointers. Prefer the slice wrappers further down.

// The raw imports are GENERATED from sdk/abi.json by tools/gen_bindings.py.
// Adding a function to the ABI does not mean editing this file; only the
// ergonomic wrappers below, and only if the new function needs one.
pub const raw = @import("raw_generated.zig");


// ============================================================= wrappers ====
//
// Slices carry their own length, so the (ptr, len) split of the wire ABI is an
// implementation detail everywhere above this line. An empty slice is passed as
// a null pointer with length 0: the host treats a zero length as the empty
// string without reading memory, but a null pointer makes that explicit rather
// than handing it Zig's dangling-but-aligned pointer for `""`.

inline fn cptr(s: []const u8) ?[*]const u8 {
    return if (s.len == 0) null else s.ptr;
}

inline fn mptr(s: []u8) ?[*]u8 {
    return if (s.len == 0) null else s.ptr;
}

// ---- Core ----------------------------------------------------------------

/// Write a line to the game log and the mod menu's log view. Messages longer
/// than 2048 bytes are truncated by the host.
pub fn log(level: LogLevel, msg: []const u8) void {
    raw.log(@intFromEnum(level), cptr(msg), @intCast(msg.len));
}

/// Fill and return the environment. Sets `size` for you, which the host needs
/// in order to know how much of your struct it may write.
pub fn env() Env {
    var e: Env = .{ .size = @sizeOf(Env) };
    raw.env(@ptrCast(&e));
    return e;
}

/// Unrecoverable error: traps out of the current call, disables the mod and
/// shows `msg` to the user. Prefer returning an error from a hook where you
/// can — a refused `mod_load` is a much better experience than a dead mod.
pub fn abortWith(msg: []const u8) noreturn {
    raw.abort(cptr(msg), @intCast(msg.len));
}

/// The instruction budget for the current hook, or `max(u64)` when unmetered.
///
/// This is the LIMIT, not a live countdown — it does not decrease as you run.
/// Size your work against it up front and count your own iterations.
pub fn fuelBudget() u64 {
    return raw.fuel_budget();
}

// ---- GameState.Read ------------------------------------------------------

pub fn turnNumber() u32 {
    return raw.turn_number();
}

pub fn countryCount() u32 {
    return raw.country_count();
}

/// `index` in [0, countryCount()). Returns INVALID if out of range. Ordering is
/// stable within a turn but not across turns.
pub fn countryAt(index: u32) Country {
    return raw.country_at(index);
}

/// The full byte length of the country's name, without writing anything.
/// This is the sizing half of the two-call pattern.
pub fn countryNameLen(c: Country) u32 {
    return raw.country_name(c, null, 0);
}

/// Fill `buf` and return the bytes actually written, which is
/// `@min(full length, buf.len)`. A name longer than `buf` is truncated, which
/// is not an error; use `countryNameLen` first if you need to know.
pub fn countryName(c: Country, buf: []u8) []u8 {
    const need = raw.country_name(c, mptr(buf), @intCast(buf.len));
    return buf[0..@min(@as(usize, need), buf.len)];
}

pub fn countryTreasury(c: Country) f64 {
    return raw.country_treasury(c);
}

pub fn countryProvinceCount(c: Country) u32 {
    return raw.country_province_count(c);
}

pub fn provincePopulation(p: Province) i64 {
    return raw.province_population(p);
}

/// Owning country, or INVALID if unowned or unknown.
pub fn provinceOwner(p: Province) Country {
    return raw.province_owner(p);
}

// ---- UI ------------------------------------------------------------------

/// Register a panel and return its handle. Returns 0 when headless, when UI was
/// revoked, or when you already hold 8 panels — none of which is fatal, so
/// degrade rather than trap. Titles are truncated to 64 bytes.
///
/// Call this from `mod_load`, not from your draw hook.
pub fn panelRegister(title: []const u8, min_w: u32, min_h: u32) Panel {
    return raw.panel_register(cptr(title), @intCast(title.len), min_w, min_h);
}

/// Filled rectangle, panel-relative. Colour is 0xRRGGBBAA. Coordinates outside
/// the panel are clipped by the host; they cannot escape it.
pub fn drawRect(panel: Panel, x: i32, y: i32, w: i32, h: i32, rgba: u32) void {
    raw.draw_rect(panel, x, y, w, h, rgba);
}

/// UTF-8 text, panel-relative. Truncated to 512 bytes per call.
pub fn drawText(panel: Panel, x: i32, y: i32, rgba: u32, text: []const u8) void {
    raw.draw_text(panel, x, y, rgba, cptr(text), @intCast(text.len));
}

/// Immediate-mode button: draws it and returns true on the frame it is clicked.
/// One click activates one button — the host consumes it, so overlapping rects
/// do not all fire. Label truncated to 64 bytes.
pub fn button(panel: Panel, x: i32, y: i32, w: i32, h: i32, label: []const u8) bool {
    return raw.button(panel, x, y, w, h, cptr(label), @intCast(label.len)) != 0;
}

// ---- Assets --------------------------------------------------------------

/// Byte size of one of your own `data/` files, or 0 if there is no such asset.
/// Names are relative to `data/` and use '/' separators: `data/flags/fr.png` is
/// `"flags/fr.png"`. The name is looked up in your package's entry list, never
/// resolved as a filesystem path.
pub fn assetSize(name: []const u8) u32 {
    return raw.size(cptr(name), @intCast(name.len));
}

/// Fill `buf` and return the bytes actually written, which is
/// `@min(asset size, buf.len)`. Two-call sizing, like `countryName`: call
/// `assetSize` first when you need the whole file.
pub fn assetRead(name: []const u8, buf: []u8) []u8 {
    const need = raw.read(cptr(name), @intCast(name.len), mptr(buf), @intCast(buf.len));
    return buf[0..@min(@as(usize, need), buf.len)];
}

// ============================================================== exports ====
//
// You write these; there is nothing here to call. Declare them in your root
// source file and pass `--export=<name>` to the linker for each one.
//
//   export fn mod_load() i32                                    required
//   export fn mod_unload() void                                 optional
//   export fn mod_pre_turn(turn: u32) void                      needs GameProcess
//   export fn mod_post_turn(turn: u32) void                     needs GameProcess
//   export fn mod_draw_panel(panel: Panel, w: u32, h: u32) void needs UI
//
// `mod_load` returns 0 to accept the load; non-zero refuses it and the value is
// shown to the user. There is no autorun, so it runs every session the user
// enables you — never assume prior state.

File: sdk/go/gearbox.go
// Package gearbox is the TinyGo binding for the OpenDoctrines mod ABI,
// Gearbox v1.0.
//
// It is a transcription of sdk/abi.json. That file is the source of truth; if
// this package disagrees with it, this package is the bug.
//
// Two directions:
//
//	IMPORTS (the raw* functions)  the host provides, your mod calls.
//	EXPORTS (mod_*)               your mod provides, the host calls.
//
// Imports live in WASM modules named "gearbox:<module>", matching the
// capability names in MANIFEST.json. You only receive imports for capabilities
// you declared AND the user granted; an undeclared import is refused at load
// time with a diagnostic, not later at an inconvenient moment.
//
// Strings on the wire are (ptr, len) pairs of UTF-8 bytes in your linear
// memory. They are NOT null-terminated. The host does not retain a pointer into
// your memory after a call returns, and you must not retain one of its.
//
// # Requires TinyGo
//
// Standard Go cannot build a mod. GOOS=js needs the JavaScript glue this host
// does not provide, and GOOS=wasip1 emits WASI imports the host deliberately
// does not link. See README.md.
//
// Status: written against sdk/abi.json, not yet compiled.
package gearbox

import "unsafe"

const (
	Major = 1
	Minor = 0
)

// Invalid is returned by any call that yields a handle it could not produce.
// Note that it is 0xFFFFFFFF, not zero — a zero handle is a real one.
const Invalid uint32 = 0xFFFFFFFF

// Opaque handles. Valid only for the duration of the hook that produced them —
// do not cache one across turns, it will be stale or reassigned. Store the
// country's name or your own key instead.
type (
	Country  uint32
	Province uint32
	Panel    uint32
)

type LogLevel uint32

const (
	LogTrace LogLevel = 0
	LogInfo  LogLevel = 1
	LogWarn  LogLevel = 2
	LogError LogLevel = 3
)

type Platform uint8

const (
	PlatformUnknown Platform = 0
	PlatformWindows Platform = 1
	PlatformMacOS   Platform = 2
	PlatformLinux   Platform = 3
	PlatformWeb     Platform = 4
)

// Env mirrors gearbox_env_t. The layout is part of the ABI: fields are only
// ever appended, never reordered or resized, and Size is what lets an older mod
// stay safe against a newer host.
//
// Use Environment() rather than filling this yourself; it sets Size, which the
// host needs in order to know how much of your struct it may write.
type Env struct {
	Size         uint32
	GearboxMajor uint32
	GearboxMinor uint32
	// HostVersion is packed: (major << 16) | (minor << 8) | patch.
	HostVersion uint32
	// Platform is a Platform value. Kept as uint8 because the host may add
	// platforms this binding has never heard of.
	Platform uint8
	// IsWeb is 1 under Emscripten. Fuel is NOT enforced there.
	IsWeb uint8
	// IsHeadless is 1 when there is no renderer. Every UI import no-ops.
	IsHeadless uint8
	Reserved0  uint8
	// ScreenW and ScreenH are 0 when headless.
	ScreenW uint32
	ScreenH uint32
}

// The four uint8 fields pack into one word and wasm32 aligns uint32 to 4, so
// this must come out at exactly the 28 bytes abi.json specifies. Either of
// these constants goes negative — and a negative uintptr constant is a compile
// error — if the size is anything else.
const _ = 28 - unsafe.Sizeof(Env{})
const _ = unsafe.Sizeof(Env{}) - 28

// ============================================================== imports ====
//
// One declaration per entry in abi.json's "imports" array. These are the wire
// ABI; prefer the wrappers below, which take Go strings and slices.
//
// The raw imports are GENERATED into raw_generated.go from sdk/abi.json by
// tools/gen_bindings.py, and live in this same package. Adding a function to
// the ABI therefore does not mean editing this file.
//
// (The module name really does contain a colon. //go:wasmimport takes the
// module and the field name as two space-separated words, and "gearbox:core"
// is one word, so it is legal even though it looks unusual.)
//
// Everything below is the ergonomic layer: Go strings to (ptr,len), two-call
// sizing, and turning INVALID into a bool-ok return. That part stays
// hand-written -- it is what makes this read as Go.

// ============================================================= wrappers ====
//
// Go strings and slices carry their own length, so the (ptr, len) split of the
// wire ABI stops at this line.
//
// An empty string or slice is passed as a nil pointer with length 0. The host
// treats a zero length as empty without reading memory either way, but the
// pointer behind a zero-length Go slice is not guaranteed to be anything, and
// handing that to a host that bounds-checks pointers is a bad habit.

func strPtr(s string) unsafe.Pointer {
	if len(s) == 0 {
		return nil
	}
	return unsafe.Pointer(unsafe.StringData(s))
}

func bufPtr(b []byte) unsafe.Pointer {
	if len(b) == 0 {
		return nil
	}
	return unsafe.Pointer(unsafe.SliceData(b))
}

// ---- Core -----------------------------------------------------------------

// Log writes a line to the game log and the mod menu's log view. Messages
// longer than 2048 bytes are truncated by the host.
func Log(level LogLevel, msg string) {
	rawLog(uint32(level), strPtr(msg), uint32(len(msg)))
}

// Environment fills and returns the environment, setting Size for you.
func Environment() Env {
	e := Env{Size: uint32(unsafe.Sizeof(Env{}))}
	rawEnv(unsafe.Pointer(&e))
	return e
}

// Abort reports an unrecoverable error: it traps out of the current call,
// disables the mod and shows msg to the user. It does not return — the trap
// happens inside the host, so nothing after the call runs, even though Go has
// no way to say so in the signature.
//
// Prefer returning non-zero from mod_load where you can. A refused load is a
// much better experience than a mod that dies mid-turn.
func Abort(msg string) {
	rawAbort(strPtr(msg), uint32(len(msg)))
}

// FuelBudget returns the instruction budget for the current hook, or
// 0xFFFFFFFFFFFFFFFF when unmetered.
//
// This is the LIMIT, not a live countdown — it does not decrease as you run.
// Size your work against it up front and count your own iterations.
func FuelBudget() uint64 {
	return rawFuelBudget()
}

// ---- GameState.Read -------------------------------------------------------

// TurnNumber returns the current turn, or 0 when no world is loaded.
func TurnNumber() uint32 { return rawTurnNumber() }

// CountryCount returns how many countries exist, or 0 when no world is loaded.
// Rebel factions are not included.
func CountryCount() uint32 { return rawCountryCount() }

// CountryAt returns the country at index in [0, CountryCount()), or Invalid if
// out of range. Ordering is stable within a turn but not across turns.
func CountryAt(index uint32) Country { return Country(rawCountryAt(index)) }

// CountryNameLen returns the full byte length of the country's name without
// writing anything. This is the sizing half of the two-call pattern.
func CountryNameLen(c Country) uint32 {
	return rawCountryName(uint32(c), nil, 0)
}

// CountryNameInto fills buf and returns the sub-slice actually written, which
// is never longer than buf. A name longer than buf is truncated, which is not
// an error; call CountryNameLen first if you need to know.
//
// This is the allocation-free form. Use it in mod_draw_panel, which runs every
// frame.
func CountryNameInto(c Country, buf []byte) []byte {
	n := rawCountryName(uint32(c), bufPtr(buf), uint32(len(buf)))
	if int(n) > len(buf) {
		n = uint32(len(buf))
	}
	return buf[:n]
}

// CountryName returns the country's name as a Go string. Convenient, and it
// allocates twice — once for the buffer, once for the string. Fine in mod_load;
// in a per-frame hook use CountryNameInto with a buffer you keep.
func CountryName(c Country) string {
	n := CountryNameLen(c)
	if n == 0 {
		return ""
	}
	buf := make([]byte, n)
	return string(CountryNameInto(c, buf))
}

// CountryTreasury returns the treasury balance, or 0 for an unknown country.
func CountryTreasury(c Country) float64 { return rawCountryTreasury(uint32(c)) }

// CountryProvinceCount returns how many provinces the country owns, or 0 for an
// unknown country.
func CountryProvinceCount(c Country) uint32 { return rawCountryProvinceCount(uint32(c)) }

// ProvincePopulation returns the population of a province, or 0 for an unknown
// province.
func ProvincePopulation(p Province) int64 { return rawProvincePopulation(uint32(p)) }

// ProvinceOwner returns the owning country, or Invalid if unowned or unknown.
func ProvinceOwner(p Province) Country { return Country(rawProvinceOwner(uint32(p))) }

// ---- UI -------------------------------------------------------------------

// PanelRegister registers a panel and returns its handle. It returns 0 when
// headless, when UI was revoked, or when you already hold 8 panels — none of
// which is fatal, so degrade rather than trap. Titles are truncated to 64
// bytes.
//
// Call this from mod_load, not from your draw hook.
func PanelRegister(title string, minW, minH uint32) Panel {
	return Panel(rawPanelRegister(strPtr(title), uint32(len(title)), minW, minH))
}

// DrawRect draws a filled rectangle in panel-relative coordinates. Colour is
// 0xRRGGBBAA. Coordinates outside the panel are clipped by the host; they
// cannot escape it.
func DrawRect(panel Panel, x, y, w, h int32, rgba uint32) {
	rawDrawRect(uint32(panel), x, y, w, h, rgba)
}

// DrawText draws UTF-8 text in panel-relative coordinates. Truncated to 512
// bytes per call.
func DrawText(panel Panel, x, y int32, rgba uint32, text string) {
	rawDrawText(uint32(panel), x, y, rgba, strPtr(text), uint32(len(text)))
}

// DrawTextBytes is DrawText for a byte slice, so a formatted line built in a
// reusable buffer does not have to be copied into a string first.
func DrawTextBytes(panel Panel, x, y int32, rgba uint32, text []byte) {
	rawDrawText(uint32(panel), x, y, rgba, bufPtr(text), uint32(len(text)))
}

// Button draws an immediate-mode button and reports whether it was clicked on
// this frame. One click activates one button — the host consumes it, so
// overlapping rects do not all fire. Label truncated to 64 bytes.
func Button(panel Panel, x, y, w, h int32, label string) bool {
	return rawButton(uint32(panel), x, y, w, h, strPtr(label), uint32(len(label))) != 0
}

// ---- Assets ---------------------------------------------------------------

// AssetSize returns the byte size of one of your own data/ files, or 0 if there
// is no such asset. Names are relative to data/ and use '/' separators:
// data/flags/fr.png is "flags/fr.png". The name is looked up in your package's
// entry list, never resolved as a filesystem path.
func AssetSize(name string) uint32 {
	return rawAssetSize(strPtr(name), uint32(len(name)))
}

// AssetReadInto fills buf and returns the sub-slice actually written, never
// longer than buf. Two-call sizing, like CountryNameInto: call AssetSize first
// when you need the whole file.
func AssetReadInto(name string, buf []byte) []byte {
	n := rawAssetRead(strPtr(name), uint32(len(name)), bufPtr(buf), uint32(len(buf)))
	if int(n) > len(buf) {
		n = uint32(len(buf))
	}
	return buf[:n]
}

// AssetRead returns a whole asset as a fresh slice, or nil if there is no such
// asset. It allocates the asset's full size — check AssetSize first if that
// might be large, because your ceiling is limits.memoryPages in the manifest
// and exceeding it traps the mod.
func AssetRead(name string) []byte {
	n := AssetSize(name)
	if n == 0 {
		return nil
	}
	return AssetReadInto(name, make([]byte, n))
}

// ============================================================== exports ====
//
// You write these; there is nothing here to call. Put them in your package main
// with //go:wasmexport (TinyGo 0.35.0+) or //export (older TinyGo):
//
//	//go:wasmexport mod_load
//	func modLoad() int32                                  required
//
//	//go:wasmexport mod_unload
//	func modUnload()                                      optional
//
//	//go:wasmexport mod_pre_turn
//	func modPreTurn(turn uint32)                          needs GameProcess
//
//	//go:wasmexport mod_post_turn
//	func modPostTurn(turn uint32)                         needs GameProcess
//
//	//go:wasmexport mod_draw_panel
//	func modDrawPanel(panel, width, height uint32)        needs UI
//
// mod_load returns 0 to accept the load; non-zero refuses it and the value is
// shown to the user. There is no autorun, so it runs every session the user
// enables you — never assume prior state.
